pythonPackages.connexion

elohmeier opened the pull request:
###### Motivation for this change
add the connexion library and dependencies to build nice Flask-API's with Nix

###### Things done
added clickclick, openapi-spec-validator, connexion


- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

worldofpeace reviewed (changes requested):
I'm seeing you've disabled tests in all of these packages because they aren't distributed from PyPI.

Usually in that case I'd recommend not fetching from PyPI and instead using the github source with `fetchFromGitHub`.

Also any dependencies needed for tests should be in `checkInputs`.

elohmeier committed:
pythonPackages.clickclick: init at 1.2.2

elohmeier committed:
pythonPackages.swagger-ui-bundle: init at 0.0.5

elohmeier committed:
pythonPackages.connexion: init at 2.3.0

elohmeier commented:
@worldofpeace thanks for your feedback! I've switched the packages to fetch from GitHub. In some cases the tests fail, so I've disabled them - but they can be executed now. I've also updated the wrong link and added another optional dependency for connexion.

worldofpeace commented on pkgs/development/python-modules/clickclick/default.nix:
```diff
@@ -0,0 +1,27 @@
+  propagatedBuildInputs = [ flake8 click pyyaml six ];
+
+  checkPhase = ''
+    py.test
```
Why not?
```suggestion
    pytest
```

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,32 @@
+  propagatedBuildInputs = [ aiohttp aiohttp-jinja2 ujson gevent jsonschema clickclick pyyaml openapi-spec-validator inflection requests flask swagger-ui-bundle ];
+
+  checkPhase = ''
+    py.test
```
Same as https://github.com/NixOS/nixpkgs/pull/52580/files#r243652803

worldofpeace commented on pkgs/development/python-modules/openapi-spec-validator/default.nix:
```diff
@@ -0,0 +1,30 @@
+  propagatedBuildInputs = [ jsonschema pyyaml six ];
+
+  checkPhase = ''
+    py.test
```
Same as https://github.com/NixOS/nixpkgs/pull/52580/files#r243652803

worldofpeace commented on pkgs/development/python-modules/swagger-ui-bundle/default.nix:
```diff
@@ -0,0 +1,25 @@
+  propagatedBuildInputs = [ jinja2 flake8 ];
+
+  # package contains no tests
+  doCheck = false;
```
:+1:

worldofpeace commented on pkgs/development/python-modules/swagger-ui-bundle/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  meta = with stdenv.lib; {
+    description = "Swagger/OpenAPI First framework on top of Flask";
+    homepage = https://github.com/zalando/connexion/;
```
I don't think this is the correct url.

worldofpeace commented on pkgs/development/python-modules/swagger-ui-bundle/default.nix:
```diff
@@ -0,0 +1,25 @@
+  pname = "swagger-ui-bundle";
+  version = "0.0.2";
+
+  src = fetchFromGitHub {
```
Is this distributed on `PyPi`?
If so we should use `fetchPypi`.

That's preferred usually, and since there's no tests anywhere that shouldn't be problem with them not having them..

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,32 @@
+  '';
+
+  # 2 upstream tests fail
+  doCheck = false;
```
Couldn't we skip those failing tests in the `pytest` invocation?

Like 
```
pytest -k 'not testname and not testname'
```

elohmeier commented:
@worldofpeace thanks for your helpful feedback! I really appreciate it =)
Thanks for the tip with pytest -k, that worked out. I've updated the packages.

worldofpeace commented on pkgs/development/python-modules/swagger-ui-bundle/default.nix:
```diff
@@ -0,0 +1,24 @@
+    sha256 = "0kk6y8p1fhmnjnxfafdw6dm1xc5wj8101y5h7hcy3rb2bxc1i94q";
+  };
+
+  propagatedBuildInputs = [ jinja2 flake8 ];
```
I think `flake8` should be a `nativeBuildInputs` because it's in `setup_requires`.

worldofpeace commented on pkgs/development/python-modules/openapi-spec-validator/default.nix:
```diff
@@ -0,0 +1,30 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    description = "OpenAPI Spec validator";
```
`description` should describe the software.
Maybe use https://github.com/p1c2u/openapi-spec-validator#about slightly.
Though it shouldn't be long or contain a line break.

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,30 @@
+  };
+
+  checkInputs = [ pytest mock flake8 pytest-aiohttp ];
+  propagatedBuildInputs = [ aiohttp aiohttp-jinja2 ujson gevent jsonschema clickclick
```
When these become long long this, it's almost better to just list them individually.

Like
```
propagatedBuildInputs = [
  aiohttp
  aiohttp-jinja2
  ujson
  gevent
  ...
];
```

If you do decide to do that, you should do similarly with `checkInputs`.

worldofpeace commented on pkgs/development/python-modules/clickclick/default.nix:
```diff
@@ -0,0 +1,28 @@
+  pname = "clickclick";
+  version = "1.2.2";
+
+  src = fetchFromGitHub {
```
Can you add a comment that we're using `fetchFromGitHub` because tests weren't distributed on PyPi?
Also do that everywhere else where this is done.

elohmeier replied:
I've added a comment.

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,70 @@
+  ]
+  ++ lib.optional (pythonOlder "3.4") pathlib
+  ++ lib.optional (pythonOlder "3.6") typing
+  ++ lib.optional isPy3k aiohttp
```
Forgot to mention `lib.optionals` which takes a list. This would make this look simpler
```suggestion
  ++ lib.optionals isPy3k [ aiohttp aiohttp-jinja2 json ]
```

I tend to use `optional` when it's singular and `optionals` when there's multiple.

elohmeier replied:
Thanks for that tip! I've changed it to "optionals".

worldofpeace commented:
@elohmeier I'm seeing that they're some deps that are already in `propagatedBuildInputs` that are needed for testing that aren't in `checkInputs` in `connexion`.

I'm not sure if that was on purpose, but they should be declared as `checkInputs` if we need them for tests. 

Also I think we're missing [`decorator`](https://github.com/zalando/connexion/blob/a0ac0fd13f6857bba2563fbdda55404cfa69dde3/setup.py#L44).

> https://github.com/NixOS/nixpkgs/pull/52580#discussion_r243776341
> I've modified that part as you suggested. However there are a lot of unit tests failing if I'm trying to build it for Python 2.7 or Python 3.6. Should we disable the tests for that versions?

Can you paste the logs somewhere after this round of review?
Perhaps we're missing something.

elohmeier committed:
pythonPackages.aiohttp-swagger: init at 1.0.5

elohmeier commented:
@worldofpeace thanks again for your review! I've been able to make the package build and test successfully now for py27, py35, py36 and py37. There was a problem using the wrong locale for Python <3.7. I've also added the missing dependecies (adding another package).
There is still a problem though, the authors of connexion need aiohttp<3.5.2. I've locally switched to 3.5.1, but nixpkgs has a more recent version on master which makes the tests fail. Do you have a suggestion on how to handle that?
Your feedback is much appreciated :-)

worldofpeace commented:
> There is still a problem though, the authors of connexion need aiohttp<3.5.2. I've locally switched to 3.5.1, but nixpkgs has a more recent version on master which makes the tests fail. Do you have a suggestion on how to handle that?
Your feedback is much appreciated :-)

Looks like they pinned it because the newer version broke their build https://github.com/zalando/connexion/pull/844.
Might want to bug them to fix that soon.

We could skip the aiohttp tests and hope it isn't broken because of it or we could pin the version within the expression.

And you could do that like
```nix
let

  aiohttp_3_5_1 = aiohttp.overridePythonAttrs (oldAttrs: {
    version = "3.5.1";

    src = fetchPypi {
      inherit (oldAttrs) pname;
      inherit version;
      sha256 = ...;
    };
  });

in
```

worldofpeace commented:
Actually you can't override it because its dependents already have `aiohttp` in their `propagatedBuildInputs` and we simply can't have that ever because it will conflict.

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,82 @@
+    sha256 = "1psbql502xa7g8q4m7mpqpdicyb2kffgq9wvrq9jr7ld22k2m282";
+  };
+
+  buildInputs = lib.optional (pythonOlder "3.7") glibcLocales;
```
Was this was only needed for the tests?

If so you can put it in `checkInputs`.

worldofpeace replied:
Also note that `checkInputs` are `nativeBuildInputs` for now.

worldofpeace commented on pkgs/development/python-modules/swagger-ui-bundle/default.nix:
```diff
@@ -0,0 +1,25 @@
+    sha256 = "0kk6y8p1fhmnjnxfafdw6dm1xc5wj8101y5h7hcy3rb2bxc1i94q";
+  };
+
+  nativeBuildInputs = [ flake8 ];
```
I'm seeing

```
setup_requires=['pytest-runner', 'flake8'],
```

Which is only useful for running tests, and we don't have any so it's just bloating the build closure.

I usually propose to patch this away.

worldofpeace commented on pkgs/development/python-modules/aiohttp-swagger/default.nix:
```diff
@@ -0,0 +1,29 @@
+buildPythonPackage rec {
+  pname = "aiohttp-swagger";
+  version = "1.0.5";
+
```
```suggestion

   # we're fetching from GitHub because tests weren't distributed on PyPi
```

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,82 @@
+    owner = "zalando";
+    repo = pname;
+    rev = version;
+    sha256 = "1psbql502xa7g8q4m7mpqpdicyb2kffgq9wvrq9jr7ld22k2m282";
```
I get a hash mismatch when I build this.

worldofpeace commented on pkgs/development/python-modules/connexion/default.nix:
```diff
@@ -0,0 +1,82 @@
+    flask
+    swagger-ui-bundle
+  ] ++ lib.optionals isPy3k [ aiohttp aiohttp-jinja2 aiohttp-swagger ujson pytest-aiohttp ];
+  propagatedBuildInputs = [
```
```suggestion
  
  propagatedBuildInputs = [
```

mmahut commented:
Are there any updates on this pull request, please?

elohmeier commented:
@worldofpeace Thanks for your feedback, I've updated my changes accordingly. I've also updated to the latest package versions.

worldofpeace commented on pkgs/development/python-modules/swagger-ui-bundle/default.nix:
```diff
@@ -0,0 +1,29 @@
+  };
+
+  # patch away unused test requirements since package contains no tests
+  patchPhase = ''
```
```suggestion
  postPatch = ''
```

elohmeier replied:
Thanks, fixed.

worldofpeace commented:
@elohmeier, we've actually undergone an API change for python packaging since this change was authored.

We have a setup hook for running pytest
* https://github.com/NixOS/nixpkgs/commit/f7e28bf5d8181926e600a222cb70180519d09726

so wherever a package uses `pytest` you should migrate like this
```
checkInputs = [
  pytestCheckHook # you should drop pytest from checkInputs as well
];

disabledTests = [
  "some_disabled_test_case"
];

elohmeier commented:
@worldofpeace Thanks for the review and info, I've switched to the new API where possible.
I've also introduced a patch to deal with the aiohttp problem from upstream, which led to no regression in the enabled tests.

worldofpeace commented on pkgs/development/python-modules/clickclick/default.nix:
```diff
@@ -0,0 +1,28 @@
+  checkInputs = [ pytestCheckHook pytestcov ];
+  propagatedBuildInputs = [ flake8 click pyyaml six ];
+
+  disabledTests = if isPy36 then [
```
```suggestion
  disabledTests = lib.optionals isPy36 [
    "test_cli"
    "test_choice_default"
  ];
```

elohmeier replied:
Thanks, fixed.

worldofpeace commented:
> @worldofpeace Thanks for the review and info, I've switched to the new API where possible.
> I've also introduced a patch to deal with the aiohttp problem from upstream, which led to no regression in the enabled tests.

Great, I'm going to build everything locally. Pretty sure we should be good to merge :+1:

worldofpeace commented:
Thanks @elohmeier :sparkles:
